File: c++/Solution739.cpp
//Stack O(n) O(n)
class Solution739 {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        vector<int> res(temperatures.size());
        stack<int> st; // 放index, 保持遞減
        for (int i = temperatures.size() - 1; i >= 0; --i) {
            while (!st.empty() && temperatures[st.top()] < temperatures[i]) {
                st.pop();
            }
            res[i] = st.empty() ? 0 : (st.top() - i);
            st.push(i);
        }
        return res;
    }
};


class Solution739_2 {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        vector<int> res(temperatures.size());
        stack<int> st;
        for (int i = 0; i < temperatures.size(); ++i) {
            while (!st.empty() && temperatures[st.top()] < temperatures[i]) {
                res[st.top()] = i - st.top(); // 由前往後就變成要一直更新之前的位置
                st.pop();
            }
            st.push(i);
        }
        return res;
    }
};

File: java/Solution739.java
package leetCode.java;

import java.util.*;

//Stack O(n) O(n)
class Solution739 {
    public int[] dailyTemperatures(int[] temperatures) {
        // 這題本質是找: 每個元素右邊第一個比它大的元素 -> Monotonic
        Deque<Integer> st = new ArrayDeque<>(); // 存index, 對應溫度遞減
        int[] res = new int[temperatures.length];
        for (int i = 0; i < temperatures.length; ++i) {
            while (!st.isEmpty() && temperatures[st.peek()] < temperatures[i]) {
                int prev = st.pop(); // 前一個比較小的溫度位置
                res[prev] = i - prev; // 計算天數間隔
            }
            st.push(i);
        }
        return res;
    }
}
/* 
index        =   0  1  2  3  4  5  6  7
temperatures = [73,74,75,71,69,72,76,73]
i=0, stack=[0]
i=1, res[0]=1-0, stack=[1]
i=2, res[1]=2-1, stack=[2]
i=3, stack=[3,2]
i=4, stack=[4,3,2]
i=5, res[4]=5-4, res[3]=5-3, stack=[5,2]
i=6, res[5]=6-5, res[2]=6-2, stack=[6]
i=7, stack=[7,6]

res=[1,1,4,2,1,1,0,0]
*/